### ISSUE TITLE
[C#] Implement new Concept Exercise: extension-methods

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1070]
This issue describes how to implement the `extension-methods` concept exercise for the C# track.

## Goal

The goal of this exercise is to teach the student the Concept of Extension Methods in C#.

## Learning objectives

- Know what extension methods are.
- Know how to define extension methods.
- Know what namespaces are.
- Know how to import namespaces.

## Out of scope

- Assemblies.

## Concepts

This Concepts Exercise's Concepts are:

- `extension-methods`: know what extension methods are; know how to define extension methods.
- `namespaces`: know what namespaces are; know how to import namespaces.

## Prequisites

This Concept Exercise's prerequisites Concepts are:

- `basics`: know how to define a static class and static method.
- `classes`: know what instance methods are.
- `numbers`: an extension method will be added to extend the `int` type.
- `strings`: an extension method will be added to extend the `string` type.

### Hints

- [Extension methods][extension-methods]: how to define extension methods.
- [Namespaces][namespaces]: how to define an import namespaces.

### After

- [Extension methods][extension-methods]: how to define extension methods.
- [Namespaces][namespaces]: how to define an import namespaces.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

If you'd like to work on implementing this exercise, the first step is to let us know through a comment on this issue, to prevent multiple people from working on the same exercise. If you have any questions while implementing the exercise, please also post them as comments in this issue.

Implementing the exercise means creating the following files:

<pre>
languages
└── csharp
    └── exercises
        └── concept
            └── extension-methods
                ├── .docs
                |   ├── after.md
                |   ├── hints.md
                |   ├── instructions.md
                |   └── introduction.md
                ├── .meta
                |   ├── config.json
                |   ├── design.md
                |   └── Example.cs
                ├── ExtensionMethods.csproj
                ├── ExtensionMethods.cs
                └── ExtensionMethodsTests.cs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should be explicit about what the student should learn from the exercise, and provide a short, concise introduction to the concept(s). The aim is to give the student just enough context to figure things out themselves and solve the exercise, as research has shown that self-discovery is the most effective learning experience. Mentioning technical terms that the student can Google if they so want, is preferable over including any code samples or an extensive description. For example we might describe a string as a "Sequence of Unicode characters" or a "series of bytes" or "an object". Unless the student needs to understand the details of what those mean to be able to solve the exercise we should not give more info in this introduction - instead allowing the student to Google, ignore, or map their existing knowledge.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

The above four files are also all described in the [concept exercises document][concept-exercises].

## Step 5: update languages/csharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "extension-methods",
        "uuid": "605a41c1-81f8-4ebd-bdad-a07b91746f8c",
        "concepts": ["extension-methods", "namespaces"],
        "prerequisites": [
          "basics",
          "classes",
          "numbers",
          "strings"
        ]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the C# track:

- `ExtensionMethods.csproj`: the C# project file.
- `ExtensionMethodsTests.cs`: the test suite.
- `ExtensionMethods.cs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `.meta/Example.cs`: an example implementation that passes all the tests.

Check out [an existing exercise][exercise-example] to see what these files should look like.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference-doc] `## Implementations` section ([example][reference-example]). As there currently is no string formatting shared document, a stub version of that file should be created at `reference/types/exception.md`.

## Step 8: updating list of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][design-example]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][config.json-example]).

### Inspiration

When implementing this exericse, it can be very useful to look at [already implemented C# exercises][implemented-exercises].

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/extension-methods.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[extension-methods]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/how-to-implement-and-call-a-custom-extension-method
[namespaces]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/namespaces/

**Comment by ryanrobidou** [comment]
Would love to take a look at implementing this exercise in the spirit of Hacktoberfest. Is this one I can help with?

**Comment by ErikSchierboom** [comment]
Yeah sure! I've updated the label. To start, I suggest the following:

- Read up on the Concept Exercise documentation https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
- If you're not already on Slack, please use this invitation to join our Slack channel: https://join.slack.com/t/exercism-team/shared_invite/enQtNDEyMDM0OTgzNzY1LTU2ZDE2MGIzZmJmMDY3MDc1M2MxMzA0YjkwNjYyY2E1ZWJmNDRjNjVhMzc4OGQzOTU3Y2JkMTQ3ZTA5MWI3OGU Then please join the #maintaining-csharp channel and also post that you're working on said issue. The C# maintainers will be able to help you.

@mikedamay Did you have any particular ideas on how you wanted to approach this exercise? I'm thinking extend one of the existing classes in the BCL.
